src/VisualStudio/Core/Next/Remote/RemotableDataJsonRpcEx.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    /// <summary>
+    /// Communication channel between remote host client and remote host.
+    /// 
+    /// this is framework's back channel to talk to remote host
+    /// 
+    /// for example, this will be used to deliver missing remotable data to remote host.
+    /// 
+    /// all connection will share one remotable data channel
+    /// </summary>
+    internal class RemotableDataJsonRpcEx : JsonRpcEx
+    {
+        private readonly IRemotableDataService _remotableDataService;
+        private readonly CancellationTokenSource _source;
+
+        /// <summary>
+        /// number of shared count
+        /// </summary>
+        private readonly object _gate;
+        private int _count;
```

Comment:
> so if someone hold onto target then everything seems fall to floor?

Yes, nothing will hold on to `T` directly.

> if there are multiple ReferenceCountedDisposables

It is an error for two `ReferenceCountedDisposable<T>` "groups" to reference the same target instance. It is impossible for this situation to occur if the `ReferenceCountedDisposable<T>` instance is created at the same time the `T` instance is created (i.e. the instance of `T` is created, and then immediately wrapped).

> how using this make things safer than the object itself implementing the reference counting?

It gets the semantics right, and is future-proofed against negative performance impact.